### PR TITLE
Improve Form::prep_value

### PR DESCRIPTION
**Pull request by kenjis**
This provides:
- specify htmlspecialchars() 2nd parameter to escape single quotes
- specify htmlspecialchars() 3rd parameter which makes more secure output. Invalid characters will be dropped

Changes:
- current fuel converts single quote to `&#39;`, but this converts it to `&#039;`
- Invalid charactors will be dropped

### CODE REVIEW

File: classes/form/instance.php [hunk 1]
```diff
@@ -573,8 +573,7 @@ public function label($label, $id = null, array $attributes = array())
 	 */
 	public function prep_value($value)
 	{
-		$value = htmlspecialchars($value);
-		$value = str_replace(array("'", '"'), array("&#39;", "&quot;"), $value);
+		$value = htmlspecialchars($value, ENT_QUOTES, \Fuel::$encoding);
```
Comment (jschreuder) [inline classes/form/instance.php]: This should just use `\Security::htmlentities($value)` as that one does exactly what you did here without code duplication.

**Comment by jschreuder** [comment]
Other than my comment good job, will merge when that's fixed.

**Comment by kenjis** [comment]
Security::htmlentities() does not escape single quotations, because the 2nd parameter of htmlentities() in it is ENT_COMPAT, not ENT_QUOTES. So if we use Security::htmlentities(), the current specification of prep_value() will be changed. Some tests I wrote fails.

In my opition, there is no clear reason not to escape single quotations. The current security practice requires to escape single quotes. So it is better to change Security::htmlentities() to use  ENT_QUOTES.

**Comment by brtriver** [comment]
I agree with kanjis.
We should not use ENT_COMPAT , but use ENT_QUOTES by default like Twig.
I want to know why ENT_COMPAT is default in Security::htmlentities().

**Comment by jschreuder** [comment]
@kenjis using `Security::htmlentities()` doesn't change the current the current security practice. `htmlentities()` uses ENT_COMPAT by default thus your changes change how it currently works. But it's not necessary to escape single quotes as all properties are enclosed in double quotes. Thus I don't see the problem here. You kind of turn it around and state "there is no clear reason not to escape single quotations" which is wrong: you should have a reason to encode, not encode just for the hell of encoding.

@brtriver I think we use ENT_COMPAT because it's secure enough for our usage and the PHP default, but the history of the Security class file doesn't go back enough to check the original commit comment.

**Comment by brtriver** [comment]
thank you for replay.

Anyone do not know whether users use single quote or double quote in HTML, so I think ENT_COMPACT is not enough.
and if use ENT_COMPACT in `htmlentities()`, it is needed to add about this in the documentation.

now:

> By default, Views use output encoding to sanitize anything you pass to them using Security::htmlentities().

... anything but single quote ...

**Comment by jschreuder** [comment]
Now that's another discussion, and not one to have in this pull-request.

> > By default, Views use output encoding to sanitize anything you pass to them using Security::htmlentities().
> > ... anything but single quote ...

Actually that's not what's meant by "anything", what's meant by that is that it sanitizes every type of variable. It says nothing about how it sanitizes. But feel free to add a clarification to the docs on this and pull-request that as well.

**Comment by kenjis** [comment]
@jschreuder current prep_value() escapes single quotes with `str_replace()`, but if we use `Security::htmlentities()`, it does not escape single quotes. So I said specification changes.

The reason to use ENT_QUOTES in `Security::htmlentities()` is just in case if HTML coders write single quotes in HTML tags.

**Comment by jschreuder** [comment]
> @jschreuder current prep_value() escapes single quotes with str_replace(), but if we use Security::htmlentities(), it does not escape single quotes. So I said specification changes.

Ah sorry missed that, hmm that does complicate this. I'd prefer to solve this right.

> The reason to use ENT_QUOTES in Security::htmlentities() is just in case if HTML coders write single quotes in HTML tags.

As I said this is another discussion outside the scope of this pull-request. We can't change this outright without offering some BC. But I may know the answer to this: I can make it a setting in config and a possible 2nd setting for Security::htmlentities() so you can force it to use ENT_QUOTES here. That'll solve both right?

**Comment by jschreuder** [comment]
Here you go: fe911cbf6eb0612b54b5fd1c0003b512845a93d4 - change it to use `Security::htmlentities($value, ENT_QUOTES);` and we're done.

Also defaults to ENT_QUOTES now but possible to overwrite in config.

**Comment by kenjis** [comment]
@jschreuder Thank you for your commit.

I've updated this pull request. 

By the way, what do you mean by "BC"?

**Comment by jschreuder** [comment]
BC = Backwards Compatibility

Thanks for the work & discussion.

**Comment by kenjis** [comment]
I see. Thanks.

**Pull request merged**